Support command 0x05 commandActiveStatusReportAlt for Fantem ZB006-X

AlexKay88 opened the pull request:
Change settings of Fantem ZB006-X device are responded by not supported command `0x05`.
Introducing support for command `0x05` by adding `commandActiveStatusReportAlt`

Find a list of detailed responses for different settings here:
https://github.com/Koenkk/zigbee2mqtt/issues/7167#issuecomment-834899178

All responses make use of:
Cluster: Unknown (`0xef00`) --> `manuSpecificTuya`
Command: Unknown (`0x05`) --> `commandActiveStatusReportAlt`

Followed @drzony description here:
https://github.com/Koenkk/zigbee2mqtt/issues/7167#issuecomment-846847132

drzony commented on src/zcl/definition/cluster.ts:
```diff
@@ -4244,6 +4244,19 @@ const Cluster: {
                     {name: 'dpValues', type: BuffaloZclDataType.LIST_TUYA_DATAPOINT_VALUES},
                 ],
             },
+            /**
```
Those are sorted by ID, so move it above ID: 6

AlexKay88 replied:
Changed order as requested.

drzony reviewed (approved):
LGTM

Koenkk commented on src/zcl/definition/cluster.ts:
```diff
@@ -4230,7 +4230,19 @@ const Cluster: {
+             * FIXME: This command is not listed in Tuya zigbee cluster description,
+             *  but there is some command 0x05 (description is: Status query)
+             *  in `Serial command list` section of the same document
+             *  So, need to investigate more information about it
```
What more needs to be investigated here?

AlexKay88 replied:
I just copied the comments from section below (ID 6) and assume it's the same for command 0x06 and 0x05.
To be honest, I did not have the time to look at the [docs](https://developer.tuya.com/en/docs/iot-device-dev/tuya-zigbee-universal-docking-access-standard?id=K9ik6zvofpzql#subtitle-6-Private%20cluster)

Keep the comment in source code or remove it?

Koenkk replied:
If it works please remove it.

Also, where is this command used? I would expect a pr to zigbee-herdsman-converters which handles this.

Koenkk replied:
I will merge them together.

AlexKay88 commented:
Thanks for merging `command 0x05 commandActiveStatusReportAlt`
In order to debug further I switched to dev branch and added on my local test system code as shown here:
https://github.com/Koenkk/zigbee-herdsman-converters/pull/4102/commits/978a06c3a23326d5778679ea725b250fd6a9be4c
https://github.com/Koenkk/zigbee-herdsman-converters/pull/4102/commits/73fd1faf2244c6a9e0f357f92bfe34f9decb22ee

Unfortunately there is nothing logged from `command 0x05 commandActiveStatusReportAlt`.
Nothing from `fromZigbee.js/ZB006X_settings` debug log lines like
`meta.logger.debug(`fromZigbee.ZB006X_settings: Found DP #${dp} with data ${JSON.stringify(dpValue)}`);`
`meta.logger.warn(`fromZigbee.ZB006X_settings: Unrecognized DP #${dp} with data ${JSON.stringify(dpValue)}`);`
and there ist no file `tuya_data_point_dump_alt` created.

I would have expected to catch frames like e.g. this one:
```
IEEE 802.15.4 Data, Dst: 0x0000, Src: 0x8553
ZigBee Network Layer Data, Dst: 0x0000, Src: 0x8553
ZigBee Application Support Layer Data, Dst Endpt: 1, Src Endpt: 1
    Frame Control Field: Data (0x00)
    Destination Endpoint: 1
    Cluster: Unknown (0xef00)
    Profile: Home Automation (0x0104)
    Source Endpoint: 1
    Counter: 14
ZigBee Cluster Library Frame
    Frame Control Field: Cluster-specific (0x09)
    Sequence Number: 114
    Command: Unknown (0x05)
Data (7 bytes)
    Data: 00016704000100
    [Length: 7]
```
(find more frames here: https://github.com/Koenkk/zigbee2mqtt/issues/7167#issuecomment-834899178 )

Any suggestions what I can do next? Is there a way to log raw frames?

AlexKay88 commented:
@drzony I added `fz.tuya_data_point_dump_alt` to the device in `devices/fantem.js` on my local filesystem.
(below code is shortened)
    {
        fingerprint: [{modelID: 'TS110F', manufacturerName: '_TZ3210_lfbz816s'}],
        model: 'ZB006-X',
        vendor: 'Fantem',
        description: 'Smart dimmer module without neutral',
        extend: extend.light_onoff_brightness({noConfigure: true}),
        fromZigbee: [...extend.light_onoff_brightness({noConfigure: true}).fromZigbee, fz.command_on, fz.command_off,
            fz.command_move, fz.command_stop, fz.tuya_data_point_dump_alt, fz.ZB006X_settings],

Is it what you meant?

drzony commented:
Yes, also make sure you are using proper zigbee-herdsman. As far as I remember you do not have proper dev environment, if possible please install proper setup on some device. It's really hard to help you when you are relying on some third party integrations to keep your code consistent.
Creating pull request and merging them to test something is not the right way to do things.
Also your implementation of tuya_data_point_dump_alt, only logs the data point and does not save anything to a file.

AlexKay88 commented:
I use a Proxmox LXC container with a clean Debian buster installation which runs mosquitto and zigbee2MQTT only.
zigbee2MQTT installation was done as described here: https://www.zigbee2mqtt.io/guide/installation/01_linux.html#installing
This container is running rock stable since >1 year now.
Recently switched to dev environment as described here: https://www.zigbee2mqtt.io/advanced/more/switch-to-dev-branch.html#linux

Is it a proper dev environment with proper zigbee-herdsman?

Only local code changes compared to current dev branch:
1) add `tuya_data_point_dump_alt` to `fromZigbee.js` (which is a slightly modified copy of `tuya_data_point_dump`:
```
    tuya_data_point_dump_alt: {
        cluster: 'manuSpecificTuya',
        type: ['commandActiveStatusReportAlt'],
        convert: (model, msg, publis, options, meta) => {
            // Don't use! This is only for tests with Fantem ZB006-X
            // tuya_data_point_dump_alt will be removed in future!
            // Changes will be integrated in tuya_data_point_dump, if tests successful
            const getHex = (value) => {
                let hex = value.toString(16);
                if (hex.length < 2) {
                    hex = '0' + hex;
                }
                return hex;
            };
            const now = Date.now().toString();
            let dataStr = '';
            for (const [i, dpValue] of msg.data.dpValues.entries()) {
                tuya.logDataPoint('tuya_data_point_dump_alt', msg, dpValue, meta);
                dataStr +=
                    now + ' ' +
                    meta.device.ieeeAddr + ' ' +
                    getHex(msg.data.seq) + ' ' +
                    getHex(i) + ' ' +
                    getHex(dpValue.dp) + ' ' +
                    getHex(dpValue.datatype);

                dpValue.data.forEach((elem) => {
                    dataStr += ' ' + getHex(elem);
                });
                dataStr += '\n';
            }
            const fs = require('fs');
            fs.appendFile('data/tuya.dump.alt.txt', dataStr, (err) => {
                if (err) throw err;
            });
        },
    },
```
2) changed type to `type: ['commandActiveStatusReportAlt'],` in `fromZigbee.js/ZB006X_settings`
```
    ZB006X_settings: {
        cluster: 'manuSpecificTuya',
        type: ['commandActiveStatusReportAlt'],
        convert: (model, msg, publish, options, meta) => {
            const dpValue = tuya.firstDpValue(msg, meta, 'ZB006X_settings');
            const dp = dpValue.dp;
            const value = tuya.getDataValue(dpValue);
            meta.logger.debug(`fz.ZB006X_settings: DP/Value/Type [${dp}|${value}|${dpValue.datatype}][${JSON.stringify(dpValue)}]`);
            if (dp === 103) {
                meta.logger.debug(`fz.ZB006X_settings.ExtSwitchType [${dp}|${value}]`);
                return {ext_switch_type: {0: 'unknown', 1: 'toggle_sw', 2: 'momentary_sw', 3: 'rotary_sw', 4: 'auto_config'}[value]};
            } else if (dp === 105) {
                meta.logger.debug(`fz.ZB006X_settings.LoadDetectionMode [${dp}|${value}]`);
                return {load_detection_mode: {0: 'none', 1: 'first_power_on', 2: 'every_power_on'}[value]};
            } else if (dp === 109) {
                meta.logger.debug(`fz.ZB006X_settings.ControlMode [${dp}|${value}]`);
                return {control_mode: {0: 'local', 1: 'remote', 2: 'both'}[value]};
            } else {
                meta.logger.warn(`fz.ZB006X_settings: Unrecognized DP [${dp}|${value}][${JSON.stringify(dpValue)}]`);
            }
        },
    },
```

3) added `fz.tuya_data_point_dump_alt` to the device in `devices/fantem.js`
```
    {
        fingerprint: [{modelID: 'TS110F', manufacturerName: '_TZ3210_lfbz816s'}],
        model: 'ZB006-X',
        vendor: 'Fantem',
        description: 'Smart dimmer module without neutral',
        extend: extend.light_onoff_brightness({noConfigure: true}),
        fromZigbee: [...extend.light_onoff_brightness({noConfigure: true}).fromZigbee, fz.command_on, fz.command_off,
            fz.command_move, fz.command_stop, fz.tuya_data_point_dump_alt, fz.ZB006X_settings],
        toZigbee: [...extend.light_onoff_brightness({noConfigure: true}).toZigbee, tz.ZB006X_settings],
        exposes: [e.light_brightness(),
            e.action(['on', 'off', 'brightness_move_down', 'brightness_move_up', 'brightness_stop']),
            exposes.enum('ext_switch_type', ea.STATE_SET, ['unknown', 'toggle_sw', 'momentary_sw', 'rotary_sw', 'auto_config'])
                .withDescription('External switch type'),
            exposes.enum('load_detection_mode', ea.STATE_SET, ['none', 'first_power_on', 'every_power_on'])
                .withDescription('Load detection mode'),
            exposes.enum('control_mode', ea.STATE_SET, ['local', 'remote', 'both']).withDescription('Control mode'),
        ],
        meta: {disableActionGroup: true},
        configure: async (device, coordinatorEndpoint, logger) => {
            await extend.light_onoff_brightness().configure(device, coordinatorEndpoint, logger);
            const endpoint = device.getEndpoint(1);
            await reporting.bind(endpoint, coordinatorEndpoint, ['genOnOff', 'genLevelCtrl']);
            await reporting.onOff(endpoint);
        },
    },
```

I hope my environment and the only changes is understandable now.
I don't get any debug logging like "fz.ZB006X_settings: Unrecognized DP",
no debug logging by `tuya.logDataPoint('tuya_data_point_dump_alt' `
and no debug log file in `data/tuya.dump.alt.txt`

drzony commented:
1. You need to do
```
cd /opt/zigbee2mqtt
git stash
git pull
git stash pop
npm ci
```
If git complains at `git stash pop` you will need to resolve conflicts otherwise your installation may be broken.
I would also advise against using the same directory for stable operation and for development.

2. If you want full environment, You also need to clone zigbee-herdsman and zigbee-herdsman-converters in their respective directories

3. You need to make sure that the proper log levels are set if you want to see all the messages

AlexKay88 commented:
Thanks for your guidance! I did it as you proposed:
```
cd /opt/zigbee2mqtt
git stash
git pull
git stash pop
npm ci
```
`git stash pop` did not raise any conflicts.

AlexKay88 commented:
I still wonder why I can't see any `cluster 0xef00/manuSpecificTuya command 0x05/ActiveStatusReportAlt` in debug logging.

A `command 0x04/sendData` is immediatley responded by a `command 0x05/ActiveStatusReportAlt` (and further `commands 0x06/ActiveStatusReport`).
Sniffer trace was taken with CC2531 (flashed with sniffer firmware) during permit join and some further device configuration done with Smart Life app while Fantem ZB006-X was connected to Tuya Zigbee Gateway. 

**Example 1 from sniffer trace:**
`Packet#525: 0x0000 --> 0x8553: cluster 0xef00/manuSpecificTuya, cmd 0x04/sendData, 00726d04000102`
which I suppose is translated to:
`00 (status) 72 (tid?) 6d (dp) 04 (dataType 4[enum]) 00 01 (value follows in next 1 bytes??) 02 (value)`
```
#527: 0x8553 --> 0x0000: cluster 0xef00/manuSpecificTuya, cmd 0x05/ActiveStatusReportAlt, 00016d04000102
00 (status) 01 (tid?) 6d (dp) 04 (dataType 4[enum]) 00 01 (value follows in next 1 bytes??) 02 (value)
```

**Example 2 from sniffer trace:**
```
#395: 0x0000 --> 0x8553: cluster 0xef00/manuSpecificTuya, cmd 0x04/sendData, 00706704000102
00 (status) 70 (tid?) 67 (data point) 04 (dataType 4[enum]) 00 01 (value follows in next 1 bytes??) 02 (value)
```
```
#396: 0x8553 --> 0x0000: cluster 0xef00/manuSpecificTuya, cmd 0x05/ActiveStatusReportAlt, 00016704000102`
00 (status) 01 (tid?) 67 (dp) 04 (dataType 4[enum]) 00 01 (value follows in next 1 bytes??) 02 (value)
```
[8 seconds later]
```
#403: 0x8553 --> 0x0000: cluster 0xef00/manuSpecificTuya, cmd 0x06/ActiveStatusReport, 00016a02000400000000
00 (status) 01 (tid?) 6a (dp) 02 (dataType 2[4 byte value]) 00 04 (value follows in next 4 bytes??) 00 00 00 00 (value)
```
[4 seconds later]
```
#410: 0x8553 --> 0x0000: cluster 0xef00/manuSpecificTuya, cmd 0x06/ActiveStatusReport, 00016a02000400000001
00 (status) 01 (tid?) 6a (dp) 02 (dataType 2[4 byte value]) 00 04 (value follows in next 4 bytes??) 00 00 00 01 (value)
```



**Pls find sniffer trace here:**
[Fantem_PermitJoin_TS110F_TZ3210_lfbz816s_0x8553_only.pcapng.gz](https://github.com/Koenkk/zigbee-herdsman/files/8530617/Fantem_PermitJoin_TS110F_TZ3210_lfbz816s_0x8553_only.pcapng.gz)
Trustcenter Link Key: `5A:69:67:42:65:65:41:6C:6C:69:61:6E:63:65:30:39`
Transport Key: `89:ab:f9:e3:68:96:5c:69:07:82:b9:e2:be:65:56:08` (useless for hackers, it's just a zigbee test gateway)

drzony commented:
Enable logging for zigbee-herdsman. You should then see the raw data. Make sure that correct version of zigbee-herdsman is installed (it should be somewhere in node_modules under /opt/zigbee2mqtt).

AlexKay88 commented:
THX, I will try to find out more!
`/opt/zigbee2mqtt/node_modules/zigbee-herdsman/package.json` says: `"version": "0.14.26",` (latest version)
Started in debug mode:
```
cd /opt/zigbee2mqtt
DEBUG=zigbee-herdsman* npm start 3>&1 1>>/opt/zigbee2mqtt/data/log/zigbee-herdsman_debug.log 2>&1 &
```

Tested again to configure device (which should send `command 0x04/sendData`) in order to expect a response with `command 0x05/ActiveStatusReportAlt`:
```
2022-04-21T18:40:28.771Z zigbee-herdsman:controller:endpoint Command 0x14b457fffe0e490b/1 manuSpecificTuya.dataRequest({"seq":35,"dpValues":[{"dp":105,"datatype":4,"data":[2]}]}, {"sendWhen":"immediate","timeout":10000,"disableResponse":false,"disableRecovery":false,"disableDefaultResponse":true,"direction":0,"srcEndpoint":null,"reservedBits":0,"manufacturerCode":null,"transactionSequenceNumber":null,"writeUndiv":false})
2022-04-21T18:40:28.771Z zigbee-herdsman:deconz:adapter no response expected
2022-04-21T18:40:28.774Z zigbee-herdsman:deconz:driver DATA_REQUEST - destAddr: 0x5e4d EP:1 SeqNr. 39 request id: 135
2022-04-21T18:40:28.802Z zigbee-herdsman:deconz:frameParser DATA_REQUEST RESPONSE - request id: 135 status: 0
2022-04-21T18:40:28.802Z zigbee-herdsman:deconz:driver networkstate: 2 apsDataConfirm: 0 apsDataIndication: 0 configChanged: 0 apsRequestFreeSlots: 1
2022-04-21T18:40:28.802Z zigbee-herdsman:deconz:frameParser DEVICE_STATE changed: 10100110
2022-04-21T18:40:28.802Z zigbee-herdsman:deconz:driver networkstate: 2 apsDataConfirm: 1 apsDataIndication: 0 configChanged: 0 apsRequestFreeSlots: 1
2022-04-21T18:40:28.806Z zigbee-herdsman:deconz:driver query aps data confirm
2022-04-21T18:40:28.835Z zigbee-herdsman:deconz:driver DATA_CONFIRM - sending data state request - SeqNr. 40
2022-04-21T18:40:28.837Z zigbee-herdsman:deconz:frameParser DATA_CONFIRM RESPONSE - destAddr: 0x5e4d request id: 135 confirm status: 0
2022-04-21T18:40:28.837Z zigbee-herdsman:deconz:driver networkstate: 2 apsDataConfirm: 0 apsDataIndication: 0 configChanged: 0 apsRequestFreeSlots: 1
2022-04-21T18:40:28.837Z zigbee-herdsman:deconz:adapter sendZclFrameToEndpoint - message send with transSeq Nr.: 226
2022-04-21T18:40:28.837Z zigbee-herdsman:deconz:adapter false, true, false
2022-04-21T18:40:28.837Z zigbee-herdsman:deconz:adapter resolve request
```
Corresponding log.txt:
```
debug 2022-04-21 20:40:28: Received MQTT message on 'zigbee2mqtt/Wohnz_Saeule/set' with data '{"load_detection_mode":"every_power_on"}'
debug 2022-04-21 20:40:28: Publishing 'set' 'load_detection_mode' to 'Wohnz_Saeule'
```

Parallel Sniffing shows that `cluster 0xef00/manuSpecificTuya command 0x00/dataRequest` is sent instead of `command 0x04/sendData`

Thus I need to change `toZigbee.js/ZB006X_settings` `tuya.sendDataPointEnum(entity, dp, value)` to `tuya.sendDataPointEnum(entity, dp, value, 4)`

Will test afterwards again...

AlexKay88 commented:
Problems solved. There is no problem in merged code.
**Solution:** Fantem ZB006-X does not use cmd 0x05 and 0x06 until [reporting of physical state changes is enabled](https://github.com/Koenkk/zigbee2mqtt/issues/9057#issuecomment-1007742130)
